### DISCUSSION TITLE
enh: folders are like projects

### DISCUSSION THREAD

**Discussion by rndmcnlly** [open-webui/open-webui#14299] [General]
Inspired the notion of Projects in Claude and the variation newly available in ChatGPT, let's think through how to give OWUI **folders** a little bit more expressive power.

# Features

## Custom instructions

Suppose that instead of just having a `Rename` options for a folder, we had a richer `Edit` option. Here are some mock-ups created quickly with the browser's devtools.

<img width="205" alt="Screenshot 2024-12-19 at 9 22 20 AM" src="https://github.com/user-attachments/assets/1c3e4249-8b9c-41b9-8b87-52b88e699965" />

<img width="432" alt="Screenshot 2024-12-19 at 9 28 19 AM" src="https://github.com/user-attachments/assets/f1df9508-bded-406a-820c-1ce9a71dc555" />

## Inheritance rules

Should the folder-specific constructions extend or override the user's general custom instructions (system prompt) or the custom instructions from an enclosing folder? I don't know. Maybe the user can decide.

To match the behavior of how the system prompt controls for a specific conversation override the user's general settings, perhaps the default value would be for the `inherit` attribute would be false. This would also match the behavior of the current ChatGPT interface where the custom instructions for a Project cause the assistant to not be aware of the user's general Custom Instructions.

## Start conversation in a folder

In order to get the immediate benefit from folder-specific custom instructions, we'd like to be able to start a new conversation in the context of a specific folder (rather than moving it into that folder after the conversation has already started). The UI for this might be a little `+` icon that is visible on hover as well as an option in the `...` menu so that the feature is available to touchscreen users.

# Use cases

I often have a few different *genres* of conversations on OWUI. Folders help me organize the conversations after they started, but I want tools to help me start them off right. Here are some examples:
- Quiet idea development: The assistant almost always replies "..." or "mhmm" to every message unless the user has a specific request. This allows stream-of-thought writing without the assistant yapping verbose replies. It is also good for voice conversations.
- Artifact construction: I want to generate in-browser artifacts, but there are bunch of special rules I want it to follow to make sure they work well within the OWUI interface.
- Song lyric generation: I want to prime the assistant to use certain formatting conventions when generating lyrics that are intended to be rendered with Suno, and I might want to give it some general background information on a repeatable songwriting process to follow in each conversation.
- Research project technical support: I want to provide background information on my current research project (including nested sub-projects) to contextualize quick text conversations that I have during meetings with my human collaborators.

# Current workarounds

## User-specific mega prompt

With one giant hunk of text in my user-specific custom instructions, I can tell the assistant the rules for each specific genre of conversation I might want. However, this slows down responses and wastes energy on irrelevant input tokens. Further, it still requires me to type out my selection of genre in my opening message rather than select it unambiguously with the mouse/touchpad.

## The *Prompts* feature in *Workspace*

If I try to put this kind of background information into prompts, it feels like the resulting conversations become cluttered with general background information at the top. Ideally, I'd like to have very short conversations where the pre-shared context information is not reproduced on the screen. Meanwhile, I don't expect to want to share many of my folder-specific instructions with others.

**Comment by flefevre** [comment]
Your proposition are very interesting.
Here some additional suggestions.

**From folder to project with team**
From my side, folder are very useful to dispatch my work but what will make more operational is to work in group .
My vision is to enable the creation of 'project' where I can add a specific knowledge, my previous threads but also add specific 'prompt' 'model assistants ' and 'human team members' .
By doing so we could be several team members both human and agentic to work in the same direction.

So introducing of projects could be pertinent as state by @rndmcnlly but what makes powerful projects is to be able to work together (humanS and AgentS) on it.

**Comment by hemzet** [comment]
If multiple Workspaces were supported in a single instance, this is the solution to the feature request. 
Our workaround is running multiple OWUI containers in the sense of "projects" or "teams". Each instance with its own team permissions, models, knowledge bases, and prompts.

**Comment by toonsevrin** [comment]
+1000 being able to organize prompts and knowledge into projects is exceptionally valuable in a professional setting! 🙏

**Comment by mansoorahmadhh** [comment]
Thats a good question. 
We are thinking because of that reasons at moment to fork OpenWebUI and make all these Enterprise Features available in a open Enterprise Solution. So everyone who is working with Enterprise thinking can help to have not only AI Interface like WordPress for normal users. Much more advanced AI Interface like TYPO3, Drupal, etc... where 1000+ employees can work in it...

- Multi Teams
- Multi Tenants Support
- Multi Workspaces
- Analytics Dashboard
- Multi Department Access
- Group Agents
- Extension, Plugin and Hooksystem to update
- AI API Keys Proxy Control with Monitoring
- Rights Managment FULLY controlled over Azure Entra ID and Google Cloud (Security)
- Out of the box Microsoft Teams Agent and Google Chat Agent Integration
- DevOPs Workspaces to allow multi Developer Teams to work on AI Apps, AI Agents and so on...
- and so on...

Means much more Features to integrate AI vertical in the companies...

Question in the round here. Its a good Idea or bad Idea?

Thanks

**Comment by hemzet** [comment]
@mansoorahmadhh 
Thanks for sharing your thoughts.

Forking OWUI brings significant challenges for future maintenance.

Additionally, the EU AI Act requires special training for businesses to handle AI by staff, hindering enterprises from implementing AI with a GUI like OWUI and opting for alternative approaches like AI workflows and agents. OWUI API also supports this, but GUI features are less necessary.

I’d focus on adding enterprise features to the main branch and review the decision and market annually.

The most important enterprise features are:
- Multi Teams
- Multi Workspaces (Multi Department Access)
- Fully controlled rights management over Azure Entra ID and Google Cloud (Security)

Cheers

**Comment by Classic298** [comment]
how does this look like to you?

https://github.com/open-webui/open-webui/pull/15194

**Reply by flefevre** [reply]
Great improvement! Will be very useful.

**Comment by Classic298** [comment]
Can be closed, is now implemented in 0.6.16

**Comment by TheMasterFX** [comment]
I saw the feature at Grok too.
<img width="710" height="370" alt="image" src="https://github.com/user-attachments/assets/f823ac39-a6fd-4b1d-aad6-54fceb20436b" />
In my understanding it doesn't modify the Systemprompt but a "pre-user-prompt". I don't know what's better.

**Comment by rndmcnlly** [comment]
This concept has now long been implemented in OWUI: https://docs.openwebui.com/features/chat-conversations/chat-features/conversation-organization#folder-settings-project-configuration
